pkg/apis/migration/migration.go

```diff
@@ -0,0 +1,351 @@
+
+var volumeMigrationServiceInitErr error
+
+// GetVolumeMigrationService returns the singleton VolumeMigrationService
+func GetVolumeMigrationService(ctx context.Context, volumeManager *cnsvolume.Manager, cnsConfig *cnsconfig.Config) (VolumeMigrationService, error) {
+	log := logger.GetLogger(ctx)
+	onceForVolumeMigrationService.Do(func() {
+		log.Info("Initializing volume migration service...")
+		volumeMigrationInstance = &volumeMigration{
+			volumePathToVolumeID: sync.Map{},
+			volumeManager:        volumeManager,
+			cnsConfig:            cnsConfig,
+		}
+		volumeMigrationServiceInitErr = k8s.CreateCustomResourceDefinition(ctx, CRDName, CRDSingular, CRDPlural,
+			reflect.TypeOf(migrationv1alpha1.CnsVSphereVolumeMigration{}).Name(), migrationv1alpha1.SchemeGroupVersion.Group, migrationv1alpha1.SchemeGroupVersion.Version, apiextensionsv1beta1.ClusterScoped)
+		if volumeMigrationServiceInitErr != nil {
+			log.Errorf("failed to create volume migration CRD. Error: %v", volumeMigrationServiceInitErr)
+			return
+		}
+		config, err := k8s.GetKubeConfig(ctx)
+		if err != nil {
+			log.Errorf("failed to get kubeconfig. err: %v", err)
+			return
+		}
+		volumeMigrationInstance.k8sClient, err = k8s.NewClientForGroup(ctx, config, CRDGroupName)
+		if err != nil {
+			log.Errorf("failed to create k8sClient. Err: %v", err)
+			return
+		}
+		go func() {
+			informer, err := k8s.GetDynamicInformer(ctx, migrationv1alpha1.SchemeGroupVersion.Group, migrationv1alpha1.SchemeGroupVersion.Version, "cnsvspherevolumemigrations", "")
+			if err != nil {
+				log.Errorf("failed to create dynamic informer for volume migration CRD. Err: %v", err)
+				return
+			}
+			handlers := cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					log.Debugf("received add event for VolumeMigration CR!")
+					var volumeMigrationObject migrationv1alpha1.CnsVSphereVolumeMigration
+					runtime.DefaultUnstructuredConverter.FromUnstructured(obj.(*unstructured.Unstructured).Object, &volumeMigrationObject)
+					volumeMigrationInstance.volumePathToVolumeID.Store(volumeMigrationObject.Spec.VolumePath, volumeMigrationObject.Spec.VolumeID)
+					log.Debugf("successfully added volumePath: %q, volumeID: %q mapping in the cache", volumeMigrationObject.Spec.VolumePath, volumeMigrationObject.Spec.VolumeID)
+				},
+				DeleteFunc: func(obj interface{}) {
+					log.Debugf("received delete event for VolumeMigration CR!")
+					var volumeMigrationObject migrationv1alpha1.CnsVSphereVolumeMigration
+					runtime.DefaultUnstructuredConverter.FromUnstructured(obj.(*unstructured.Unstructured).Object, &volumeMigrationObject)
+					volumeMigrationInstance.volumePathToVolumeID.Delete(volumeMigrationObject.Spec.VolumePath)
+					log.Debugf("successfully deleted volumePath: %q, volumeID: %q mapping from cache", volumeMigrationObject.Spec.VolumePath, volumeMigrationObject.Spec.VolumeID)
+				},
+			}
+			informer.Informer().AddEventHandler(handlers)
+			stopCh := make(chan struct{})
```

Comment:
Any reason why we are not using stopCh to stop the informer?

Reply:
when do we need to stop? We need to keep informer running lifetime of the container.